Actually support user-defined `op` in `rasterize`

asinghvi17 opened the pull request:
Even with providing `eltype` and `init` there were certain things trying to be computed that shouldn't have.  This fixes that. 

Also adds a test about this and streamlines CI + adds a `lts` job, as well as a doc example


Here's the MWE from slack:
```julia
using Rasters, ArchGDAL
import GeoInterface as GI
using GADM
countries_fc = GADM.get("USA"; depth = 1)

ext = Rasters.Extent(X = (-180, 180), Y = (-90, 90))
width, height = 720, 360

target = Rasters.create(
    Vector{Int},
    ext;
    size = (width, height),
    fill = Int[],
    missingval = Int[],
)

rasterize!(
    target,
    countries_fc;
    fill = [[i] for i in 1:GI.nfeature(countries_fc)],
    missingval = Int[],
    init = Int[],
    op = vcat,
    boundary = :touches,
    eltype = Vector{Int},
    progress = true,
)
```

There seems to be another issue here though, that apparently no two polygons in the dataset share the same cell?!  Which seems...incorrect

asinghvi17 committed:
Actually respect user's preference of `op`

How have we not come across this before?

asinghvi17 committed:
Only get `zero(filleltype)` if eltype is not known

tiemvanderdeure commented on src/methods/rasterize.jl:
```diff
@@ -120,8 +120,12 @@ function Rasterizer(geom, fill, fillitr;
+        _reduce_op(reducer)
+    else
+        op
+    end
```
This actually fixes the problem with op becoming `nothing`. Will need a test!

asinghvi17 committed:
if statement -> ternary

Co-authored-by: Tiem van der Deure <tiemvanderdeure@gmail.com>

asinghvi17 committed:
Add a test for the `vcat` op

With a vector of Ints and a specified eltype.  Just to make sure this pipeline works.

asinghvi17 commented:
Ok so apparently `rasterize!` gives me a different result than `rasterize`.  That's also not good

asinghvi17 commented:
Here's the usecase MWE:

```
using Rasters, ArchGDAL
import GeoInterface as GI
using GADM
countries_fc = GADM.get("USA"; depth = 1)

ext = Rasters.Extent(X = (-180, 180), Y = (-90, 90))
width, height = 720, 360

target = Rasters.create(
    Vector{Int},
    ext;
    size = (width, height),
    fill = Int[],
    missingval = Int[],
)

bang = rasterize!(
    target,
    countries_fc;
    fill = [[i] for i in 1:GI.nfeature(countries_fc)],
    missingval = Int[],
    init = Int[],
    op = vcat,
    boundary = :touches,
    eltype = Vector{Int},
    progress = true,
)

nobang = rasterize(
    countries_fc;
    fill = [[i] for i in 1:GI.nfeature(countries_fc)],
    missingval = Int[],
    init = Int[],
    op = vcat,
    boundary = :touches,
    eltype = Vector{Int},
    size = (width, height),
    to = ext,
    progress = true,
)

unique(bang)
unique(nobang)

```

asinghvi17 committed:
Add LTS CI

asinghvi17 committed:
use modern CI header from template

asinghvi17 committed:
Cheeky crazy rasterization tutorial

asinghvi17 committed:
Delete bad theme reference

We need to clean up that theme, apparently

asinghvi17 committed:
fix plotting tutorial by namespacing

rafaqz commented:
Lets put these point/interval discussion somewhere else it will get lost when I merge this

rafaqz commented on docs/src/tutorials/crazy_rasterization.md:
```diff
@@ -0,0 +1,54 @@
+ras = rasterize(
+    countries;
+    op = vcat, 
+    fill = [[i] for i in 1:length(countries.geometry)],
```
```suggestion
    fill = [[i] for i in 1:length(countries.geometry)], # By default a Vector with just the current country id
```

asinghvi17 committed:
add a comment about what the fill expression is doing

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz commented on docs/src/tutorials/crazy_rasterization.md:
```diff
@@ -0,0 +1,54 @@
+where two countries overlap, you want to save the indices of both countries.
+
+A simple but inefficient way to do this is to have each pixel of the raster actually be an
+array of integers, which stores the index of each geometry that touches the pixel.  Then you
```
```suggestion
array of integers, which stores the index of each geometry that touches the pixel.
```

asinghvi17 committed:
Update docs/src/tutorials/crazy_rasterization.md

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz commented on docs/src/tutorials/crazy_rasterization.md:
```diff
@@ -0,0 +1,54 @@
+
+A simple but inefficient way to do this is to have each pixel of the raster actually be an
+array of integers, which stores the index of each geometry that touches the pixel.  Then you
+can control exactly what you want to do.
```
```suggestion
Then we can use `rasterize` to add or remove countries, here using `vcat` to simply join merge vectors of country numbers.
```

asinghvi17 committed:
Update docs/src/tutorials/crazy_rasterization.md

Co-authored-by: Rafael Schouten <rafaelschouten@gmail.com>

rafaqz commented on docs/src/tutorials/crazy_rasterization.md:
```diff
@@ -0,0 +1,54 @@
+Let's take an example.  Say you want to rasterize all the countries in the world, but 
+where two countries overlap, you want to save the indices of both countries.
+
+A simple but inefficient way to do this is to have each pixel of the raster actually be an
```
We are just showing that you can hack, not the perfect way to hack

rafaqz committed:
Update docs/src/tutorials/crazy_rasterization.md

asinghvi17 committed:
More plotting tutorial fixes

asinghvi17 commented:
Hmm `sample` tests are failing on `convert`.   Seems unrelated to this PR but maybe that should be `GI.convert`?

asinghvi17 committed:
Fix Documenter concurrency group in CI

tiemvanderdeure commented:
I think it's a great idea to add a rasterization tutorial - but I would prefer if it started with some more normal example and then we can have a crazy rasterization thing in the end to really show off what is possible. But since we don't have the common use cases documented as a tutorial it seems weird to add a tutorial with something so niche alone

rafaqz commented:
Do I need to fix this bug?

asinghvi17 committed:
Deploy docs even if there are errors

asinghvi17 committed:
print correctly

tiemvanderdeure commented:
> I have no clue why `sample` might be failing - didn't touch that code! And a brief look shows it doesn't seem to rely on `rasterize` at all.

It must have something to do with the changes to CI you made. I've just run Rasters test locally and no issues. My guess is we hvae  a bug specific to the lts

asinghvi17 committed:
enclose bad video in comments

asinghvi17 committed:
Don't fail fast in CI

asinghvi17 commented:
That seems a bit odd.  Let's see what tests on v1 give us, though they'll probably fail because of the 1.12 issue with threadid...

But I guess we just haven't been tracking lts, the CI setup here hadn't been updated in a couple years.  I also stopped the fail-fast thing, it doesn't really help when tracking down julia version issues

asinghvi17 committed:
CI on 1.11 too

tiemvanderdeure commented:
So looking at the tests, 1.12 fails on threadid as expected, 1.11 fails on a floating point issue in the reproject tests that I already fixed in #1030, and lts fails on sample - which we should investigate.

asinghvi17 commented:
If #1030 is working can we get it in now and then use that as a basis for the rest of the PRs?  That way, at least CI works.  cc @rafaqz 

We can also move the CI changes from here to there

tiemvanderdeure committed:
Merge branch 'main' into as/flexible-rasterize

rafaqz committed:
Merge branch 'main' into as/flexible-rasterize

tiemvanderdeure commented:
> Ok this is very weird...1.12 ci is passing but 1.11 and lts are not?!

Not all that weird - since previously we only tested on 1.12 and made tests pass their.

The 1.11 issue is this test https://github.com/rafaqz/Rasters.jl/blob/bf04129a9c32eb9b088fbc3bcf6d526cffd00969/test/sources/gdal.jl#L46

Reproduced it like this:
```julia
include(joinpath(dirname(pathof(Rasters)), "../test/test_utils.jl"))
url = "https://download.osgeo.org/geotiff/samples/gdal_eg/cea.tif"
gdalpath = maybedownload(url)
Raster(gdalpath; missingval=missing) |> parent |> typeof
```

On 1.11 this is `Base.ReshapedArray{Union{Missing, UInt8}, 2, Array{Union{Missing, UInt8}, 3}, Tuple{}}` and on 1.12 it is `Matrix{Union{Missing, UInt8}}`. I have no clue why though. I "fixed" this test to make it pass - but maybe we should just test that the parent is an `AbstractArray{Union{Missing, UInt8}}`

On 1.10 it fails because of this sample issue which we'll have to look into.

asinghvi17 committed:
Fix 1.11 CI by checking only for AbstractMatrix-ness

I guess 1.12 is able to drop the wrapper and the extra dim but 1.11 is not

tiemvanderdeure committed:
Fix Rasters.sample on LTS (#1035)

Co-authored-by: Anshul Singhvi <anshulsinghvi@gmail.com>
